File: ruby/binary-search-tree/binary_search_tree.rb
class Bst
  attr_reader :data, :left, :right

  def initialize(data)
    @data = data
  end

  def insert(number)
    number <= @data ? insert_left(number) : insert_right(number)
  end

  def each(&block)
    return to_enum unless block_given?
    
    @left.each(&block) if @left
    yield data
    @right.each(&block) if @right
  end

  def insert_left(number)
    if @left
      @left.insert(number)
    else
      @left = Bst.new(number)
    end
  end

  def insert_right(number)
    if @right
      @right.insert(number)
    else
      @right = Bst.new(number)
    end
  end
end

module BookKeeping
  VERSION = 1
end

File: JavaScript/binary-search-tree/binary-search-tree.js
export class BinarySearchTree {
    constructor(data) {
        this.data = data;
        this.left;
        this.right;
    }

    each(callback) {
        if (this.left) {
            this.left.each(callback);
        }
        callback(this.data);
        if (this.right) {
            this.right.each(callback);
        }
    }

    insert(number) {
        number <= this.data ? this.insert_left(number) : this.insert_right(number);
    }

    insert_left(number) {
        if (this.left) {
            this.left.insert(number);
        } else {
            this.left = new BinarySearchTree(number);
        }
    }

    insert_right(number) {
        if (this.right) {
            this.right.insert(number);
        } else {
            this.right = new BinarySearchTree(number);
        }
    }
}